src/humanloop/otel/exporter.py

```diff
@@ -0,0 +1,305 @@
+import copy
+import json
+import logging
+import typing
+from queue import Empty as EmptyQueue
+from queue import Queue
+from threading import Thread
+from typing import Any, Optional
+
+from opentelemetry import trace
+from opentelemetry.sdk.trace import ReadableSpan
+from opentelemetry.sdk.trace.export import SpanExporter, SpanExportResult
+
+from humanloop.core import ApiError as HumanloopApiError
+from humanloop.eval_utils.context import EvaluationContext
+from humanloop.otel import TRACE_FLOW_CONTEXT, FlowContext
+from humanloop.otel.constants import HUMANLOOP_FILE_KEY, HUMANLOOP_FILE_TYPE_KEY, HUMANLOOP_LOG_KEY, HUMANLOOP_PATH_KEY
+from humanloop.otel.helpers import is_humanloop_span, read_from_opentelemetry_span
+from humanloop.requests.flow_kernel_request import FlowKernelRequestParams
+from humanloop.requests.prompt_kernel_request import PromptKernelRequestParams
+
+if typing.TYPE_CHECKING:
+    from humanloop.client import Humanloop
+
+
+logger = logging.getLogger("humanloop.sdk")
+
+
+class HumanloopSpanExporter(SpanExporter):
+    """Upload Spans created by SDK decorators to Humanloop.
+
+    Spans not created by Humanloop SDK decorators will be ignored.
+    """
+
+    DEFAULT_NUMBER_THREADS = 4
+
+    def __init__(
+        self,
+        client: "Humanloop",
+        worker_threads: Optional[int] = None,
+    ) -> None:
+        """Upload Spans created by SDK decorators to Humanloop.
+
+        Spans not created by Humanloop SDK decorators will be ignored.
+        """
+        super().__init__()
+        self._client = client
+        # Uploaded spans translate to a Log on Humanloop. The IDs are required to link Logs in a Flow Trace
+        self._span_id_to_uploaded_log_id: dict[int, Optional[str]] = {}
+        # Work queue for the threads uploading the spans
+        self._upload_queue: Queue = Queue()
+        # Worker threads to export the spans
+        self._threads: list[Thread] = [
+            Thread(
+                target=self._do_work,
+                daemon=True,
+            )
+            for _ in range(worker_threads or self.DEFAULT_NUMBER_THREADS)
+        ]
+        # Signals threads no more work will arrive and
```

Comment:
i'd expect more `logger.info` and `logger.debug` statements.

Reply:
(throughout this file)

Reply:
ack added more debug statements for threaded sections